sharktank/sharktank/ops/utils.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2025 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import functools
+
+from typing import Callable, Sequence, Any
+from sharktank.utils import tree
+
+
+def call_trivially_replicable(
+    fn: Callable[..., Any], args: tuple[Any, ...], kwargs: dict[str, Any]
+) -> Any:
+    """Call an function with replicated tensor arguments that can be called with
+    unsharded tensors.
+
+    It is expected that all replicated tensors have the same number of shards.
+    The function will be called with each set of matching shard indices and
+    replicated tensor(s) will be constructed from the resulting shards.
+
+    This wrapper handles tee-like structures in the arguments and results. It will
+    traverse them and handle nested tensors. Note that sequence and dictionary types
+    will be reconstructed with dict and tuple types, so the underlying function fn
+    must be polymorphic in this regard and not expect concrete types.
+
+    When collecting the results all non-tensor results are taken from the first device.
+    """
+
+    from sharktank.types import (
+        ReplicatedTensor,
+        is_any_tensor,
+    )
+
+    def is_leaf(v: Any) -> bool:
+        return is_any_tensor(v) or tree.is_leaf_default(v)
+
+    flat_args = tree.flatten(
+        (
+            args,
+            kwargs,
+        ),
+        is_leaf=is_leaf,
+    )
+    first_replicated_tensor_arg = [
+        arg for arg in flat_args if isinstance(arg, ReplicatedTensor)
+    ]
+    first_replicated_tensor_arg = (
+        None
+        if len(first_replicated_tensor_arg) == 0
+        else first_replicated_tensor_arg[0]
+    )
+    if first_replicated_tensor_arg is None:
+        # No replicated tensor arguments, just call the function.
+        return fn(*args, **kwargs)
+
+    shard_count = first_replicated_tensor_arg.shard_count
+
+    per_shard_args_and_kwargs = [
+        _extract_per_shard_args_and_kwargs(i, args, kwargs) for i in range(shard_count)
+    ]
+    per_shard_results = [
+        fn(*per_shard_args, **per_shard_kwargs)
+        for per_shard_args, per_shard_kwargs in per_shard_args_and_kwargs
+    ]
+
+    def reduce_to_list_of_tensors_fn(a: Any, b: Any) -> Any:
+        if not is_any_tensor(b):
+            return a
+        a.append(b)
+        return a
+
+    def tensor_to_empty_list(x: Any) -> list:
```

Comment:
The type hint for the return isn't correct. It should be either `list | Any` or just `Any`.

Reply:
Done.